src/util/semaphore.js

```diff
@@ -65,6 +65,40 @@ class Semaphore {
         }
     }
 
+    /**
+     * submit function without blocking only if there  is enough count.
+     * @param {*} count
+     * @param {*} func
+     */
+    submit_background(count, func) {
+        if (!this.try_acquire_nonblocking(count)) return;
+
+        setImmediate(async () => {
+            try {
+                await func();
+            } finally {
+                this.release(count);
+            }
+        });
+    }
+
+    /**
+     * opportunistic acquire, only if no waiters and enough value.
+     * @returns {boolean} true only if we actually managed to acquire the count from the semaphore value.
+     */
+    try_acquire_nonblocking(count) {
```

Comment:
Lets just dedup that code from wait() below:
```javascript

    async wait(count) {
        count = to_sem_count(count);

        if (this.try_acquire_nonblocking(count)) return;

/***** continue as is ******/

        if (this._timeout && !this._timer) {
            this._timer = setTimeout(
                () => this._on_timeout(),
                this._timeout
            );
        }

        this._waiting_value += count;

        // push the waiter's count to the wait queue and return a promise
        const waiter = {
            count: count,
            time: Date.now(),
        };

        return this._wq.wait(waiter);
    }
```

Reply:
ok